### ISSUE TITLE
plot_saga_vs_svrg.py broken

### ISSUE THREAD

**Issue by fabianp** [openopt/copt#56]
with numba installed, I get the error
"""
File "copt/utils.py", line 379:
        def log_deriv(p, y):
            <source elided>

            tmp[p > 0] = np.exp(-p[p > 0])
            ^

[1] During: resolving callee type: type(CPUDispatcher(<function LogLoss.partial_deriv.<locals>.log_deriv at 0x7f9d81486710>))
[2] During: typing of call at /home/pedregosa/dev/copt/copt/randomized.py (179)

[3] During: resolving callee type: type(CPUDispatcher(<function LogLoss.partial_deriv.<locals>.log_deriv at 0x7f9d81486710>))
[4] During: typing of call at /home/pedregosa/dev/copt/copt/randomized.py (179)

File "copt/randomized.py", line 179:
    def _saga_epoch(x, idx, memory_gradient, gradient_average, grad_tmp, step_size):
        <source elided>
                p += x[j_idx] * A_data[j]
            grad_i = f_deriv(p, b[i])
"""

Regression seems to be introduced in 28730398

**Comment by fabianp** [comment]
CC @GeoffNN because of 2873039805c690a95cc7f64683f5de84deb1f4ac

**Comment by GeoffNN** [comment]
I've reproduced the error, getting the following message:

```
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of Function(<built-in function getitem>) with argument(s) of type(s): (float64, bool)
 * parameterized
In definition 0:
    All templates rejected with literals.
In definition 1:
    All templates rejected without literals.
In definition 2:
    All templates rejected with literals.
In definition 3:
    All templates rejected without literals.
In definition 4:
    All templates rejected with literals.
In definition 5:
    All templates rejected without literals.
In definition 6:
    All templates rejected with literals.
In definition 7:
    All templates rejected without literals.
In definition 8:
    All templates rejected with literals.
In definition 9:
    All templates rejected without literals.
In definition 10:
    All templates rejected with literals.
In definition 11:
    All templates rejected without literals.
In definition 12:
    All templates rejected with literals.
In definition 13:
    All templates rejected without literals.
In definition 14:
    All templates rejected with literals.
In definition 15:
    All templates rejected without literals.
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: typing of intrinsic-call at /home/geoff/PycharmProjects/copt/copt/utils.py (379)
File "../copt/utils.py", line 379:
        def log_deriv(p, y):
            <source elided>
            tmp[p > 0] = np.exp(-p[p > 0])
            ^
[1] During: resolving callee type: type(CPUDispatcher(<function LogLoss.partial_deriv.<locals>.log_deriv at 0x7f95c924e0e0>))
[2] During: typing of call at /home/geoff/PycharmProjects/copt/copt/randomized.py (179)
[3] During: resolving callee type: type(CPUDispatcher(<function LogLoss.partial_deriv.<locals>.log_deriv at 0x7f95c924e0e0>))
[4] During: typing of call at /home/geoff/PycharmProjects/copt/copt/randomized.py (179)
File "../copt/randomized.py", line 179:
    def _saga_epoch(x, idx, memory_gradient, gradient_average, grad_tmp, step_size):
        <source elided>
                p += x[j_idx] * A_data[j]
            grad_i = f_deriv(p, b[i])
            ^
```

**Comment by GeoffNN** [comment]
This happens when `log_deriv` is called with a `float` instead of an `numpy.ndarray`. I introduced this in 2873039805c690a95cc7f64683f5de84deb1f4ac to handle `batch_size > 1`.

**Comment by fabianp** [comment]
An easy fix might be to make sure SAGA / SVRG call it with an array

**Comment by fabianp** [comment]
Those methods were not designed to use minibatches, so this might be a good opportunity to improve that ;-)

**Comment by GeoffNN** [comment]
This issue was fixed by #60.
I'll start a new issue for mini-batches as a To Do.
